Replace openssl hash functions with RustCrypto hashes

Fethbita opened the pull request:
Aim is to remove openssl dependency eventually. See https://github.com/RustCrypto/traits/issues/1683 for more details on where I am stuck right now.

Fethbita committed:
WIP RustCrypto hash functions addition

See https://github.com/RustCrypto/traits/issues/1683 for more details on
errors

Fethbita committed:
Remove ecb package and rename Boring error to openssl in tests

Fix use_digestalg test

Fethbita committed:
Merge branch 'main' into feat-rustcrypto

dishmaker commented on src/lib.rs:
```diff
@@ -1168,43 +1171,55 @@ fn des3_adjust_parity_bits(mut key: Vec<u8>) -> Vec<u8> {
+    } else if Oid::const_new(&[2, 16, 840, 1, 101, 3, 4, 2, 2]).eq(oid) {
+        Ok(Box::new(sha2::Sha384::default()))
+    } else if Oid::const_new(&[2, 16, 840, 1, 101, 3, 4, 2, 1]).eq(oid) {
+        Ok(Box::new(sha2::Sha256::default()))
```
There might be a better way. But probably slower
https://github.com/RustCrypto/formats/blob/0bd9dea37e3d8cccdb5f7ba5a402f62553d07e05/cms/src/builder.rs#L1061-L1077

Fethbita replied:
Hey, thanks for the comment. It seems that with that approach, I would need to use an oid database (`der::oid::db::DB`). Since oids do not change, I think it's better to have them hardcoded. Even though it looks cleaner, I would rather have very little dependency footprint. Currently, I have quite many dependencies but I plan to reduce them in time.

Fethbita committed:
Merge branch 'main' into feat-rustcrypto

Fethbita committed:
WIP implementation of certificate verification

Fethbita committed:
Add more of signature verification

RustCrypto is missing brainpool curves, so the signature verification is not implemented still

RSA and DSA are also not implemented

Fethbita commented:
RustCrypto is missing Brainpool and some other elliptic curve algorithms (https://github.com/RustCrypto/elliptic-curves/issues/114)

Fethbita commented on src/lib.rs:
```diff
@@ -1830,6 +2227,238 @@ pub fn parse_master_list(master_list: &[u8]) -> Result<X509Store, EmrtdError> {
+    // <https://www.icao.int/publications/Documents/9303_p12_cons_en.pdf>
+    // > MUST be v3
+    if cert.tbs_certificate.version.ne(&rasn_pkix::Version::V3) {
+        error!("DSC TBSCertificate version must be v3");
```
s/DSC/certificate

Fethbita commented on src/lib.rs:
```diff
@@ -2287,24 +2908,27 @@ pub fn passive_authentication(
-        ));
+    let signature_algorithm = &signer_info.signature_algorithm;
+    let signature = der::encode(&signer_info.signature).map_err(EmrtdError::RasnEncodeError)?;
+    // let pub_key = dsc.public_key().map_err(EmrtdError::OpensslErrorStack)?;
```
Remove commented code

Fethbita commented on src/lib.rs:
```diff
@@ -1776,27 +2172,28 @@ pub fn parse_master_list(master_list: &[u8]) -> Result<X509Store, EmrtdError> {
-        return Err(EmrtdError::VerifySignatureError(
-            "Signature verification failure during Master List parsing",
-        ));
+    // let pub_key = master_list_signer
```
Remove commented code

Fethbita commented on src/lib.rs:
```diff
@@ -1434,107 +1835,103 @@ pub fn parse_master_list(master_list: &[u8]) -> Result<X509Store, EmrtdError> {
+                        // let chain = Stack::new().map_err(EmrtdError::OpensslErrorStack)?;
+                        // let mut store_bldr =
+                        //     X509StoreBuilder::new().map_err(EmrtdError::OpensslErrorStack)?;
+                        // store_bldr
```
Remove commented code

Fethbita commented on src/lib.rs:
```diff
@@ -1830,6 +2227,238 @@ pub fn parse_master_list(master_list: &[u8]) -> Result<X509Store, EmrtdError> {
+        // <https://www.icao.int/publications/Documents/9303_p12_cons_en.pdf>
+        // > Dates through 2049 MUST be in UTCTime UTCTime MUST be represented as YYMMDDHHMMSSZ
+        rasn_pkix::Time::Utc(date_time) => {
+            // todo!()
```
easy todos

dishmaker commented:
Looks like only `bp512` is missing 😄 
https://github.com/RustCrypto/elliptic-curves

Fethbita commented:
> Looks like only `bp512` is missing 😄 https://github.com/RustCrypto/elliptic-curves

Oh, indeed. That one is still needed for the Master List check.
